// vue-sonner/lib/vue-sonner.js
var me = Object.defineProperty;
var pe = (s, t, e) => t in s ? me(s, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : s[t] = e;
var y = (s, t, e) => (pe(s, typeof t != "symbol" ? t + "" : t, e), e);
import "./index.css";
import { defineComponent as ee, openBlock as l, createElementBlock as h, createElementVNode as $, Fragment as K, renderList as ae, unref as M, ref as c, computed as w, onMounted as Q, watchEffect as q, onUnmounted as re, normalizeClass as se, normalizeStyle as X, createVNode as ge, createCommentVNode as C, createBlock as O, createTextVNode as oe, toDisplayString as z, resolveDynamicComponent as ye, useAttrs as we, nextTick as be, watch as Te } from "vue";
let Z = 1;
class xe {
  constructor() {
    y(this, "subscribers");
    y(this, "toasts");
    // We use arrow functions to maintain the correct `this` reference
    y(this, "subscribe", (t) => (this.subscribers.push(t), () => {
      const e = this.subscribers.indexOf(t);
      this.subscribers.splice(e, 1);
    }));
    y(this, "publish", (t) => {
      this.subscribers.forEach((e) => e(t));
    });
    y(this, "addToast", (t) => {
      this.publish(t), this.toasts = [...this.toasts, t];
    });
    y(this, "create", (t) => {
      var p;
      const { message: e, ...i } = t, u = typeof (t == null ? void 0 : t.id) == "number" || ((p = t.id) == null ? void 0 : p.length) > 0 ? t.id : Z++, m = this.toasts.findIndex((v) => v.id === u), d = t.dismissible === void 0 ? !0 : t.dismissible;
      return m !== -1 ? this.toasts = this.toasts.map((v) => v.id === u ? (this.publish({ ...v, ...t, id: u, title: e }), {
        ...v,
        ...t,
        id: u,
        dismissible: d,
        title: e
      }) : v) : this.addToast({ title: e, ...i, dismissible: d, id: u }), u;
    });
    y(this, "dismiss", (t) => (t || this.toasts.forEach((e) => {
      this.subscribers.forEach(
        (i) => i({ id: e.id, dismiss: !0 })
      );
    }), this.subscribers.forEach((e) => e({ id: t, dismiss: !0 })), t));
    y(this, "message", (t, e) => this.create({ ...e, message: t }));
    y(this, "error", (t, e) => this.create({ ...e, type: "error", message: t }));
    y(this, "success", (t, e) => this.create({ ...e, type: "success", message: t }));
    y(this, "info", (t, e) => this.create({ ...e, type: "info", message: t }));
    y(this, "warning", (t, e) => this.create({ ...e, type: "warning", message: t }));
    y(this, "loading", (t, e) => this.create({ ...e, type: "loading", message: t }));
    y(this, "promise", (t, e) => {
      if (!e)
        return;
      let i;
      e.loading !== void 0 && (i = this.create({
        ...e,
        promise: t,
        type: "loading",
        message: e.loading
      }));
      const u = t instanceof Promise ? t : t();
      let m = i !== void 0;
      return u.then((d) => {
        if (e.success !== void 0) {
          m = !1;
          const p = typeof e.success == "function" ? e.success(d) : e.success;
          this.create({ id: i, type: "success", message: p });
        }
      }).catch((d) => {
        if (e.error !== void 0) {
          m = !1;
          const p = typeof e.error == "function" ? e.error(d) : e.error;
          this.create({ id: i, type: "error", message: p });
        }
      }).finally(() => {
        var d;
        m && (this.dismiss(i), i = void 0), (d = e.finally) == null || d.call(e);
      }), i;
    });
    // We can't provide the toast we just created as a prop as we didn't create it yet, so we can create a default toast object, I just don't know how to use function in argument when calling()?
    y(this, "custom", (t, e) => {
      const i = (e == null ? void 0 : e.id) || Z++;
      return this.publish({ ...e, id: i, title: t }), i;
    });
    this.subscribers = [], this.toasts = [];
  }
}
const T = new xe(), ke = (s, t) => {
  const e = (t == null ? void 0 : t.id) || Z++;
  return T.publish({
    title: s,
    ...t,
    id: e
  }), e;
}, $e = ke, $t = Object.assign($e, {
  success: T.success,
  info: T.info,
  warning: T.warning,
  error: T.error,
  custom: T.custom,
  message: T.message,
  promise: T.promise,
  dismiss: T.dismiss,
  loading: T.loading
}), Be = ["data-visible"], Ce = { className: "sonner-spinner" }, _e = /* @__PURE__ */ ee({
  __name: "Loader",
  props: {
    visible: Boolean
  },
  setup(s) {
    const t = Array(12).fill(0);
    return (e, i) => (l(), h("div", {
      className: "sonner-loading-wrapper",
      "data-visible": s.visible
    }, [
      $("div", Ce, [
        (l(!0), h(K, null, ae(M(t), (u) => (l(), h("div", {
          key: `spinner-bar-${u}`,
          className: "sonner-loading-bar"
        }))), 128))
      ])
    ], 8, Be));
  }
}), H = (s, t) => {
  const e = s.__vccOpts || s;
  for (const [i, u] of t)
    e[i] = u;
  return e;
}, Ee = {}, Ie = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 20 20",
  fill: "currentColor",
  height: "20",
  width: "20"
}, Ne = /* @__PURE__ */ $("path", {
  "fill-rule": "evenodd",
  d: "M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z",
  "clip-rule": "evenodd"
}, null, -1), Oe = [
  Ne
];
function Se(s, t) {
  return l(), h("svg", Ie, Oe);
}
const Me = /* @__PURE__ */ H(Ee, [["render", Se]]), Pe = {}, Ae = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 20 20",
  fill: "currentColor",
  height: "20",
  width: "20"
}, De = /* @__PURE__ */ $("path", {
  "fill-rule": "evenodd",
  d: "M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7-4a1 1 0 11-2 0 1 1 0 012 0zM9 9a.75.75 0 000 1.5h.253a.25.25 0 01.244.304l-.459 2.066A1.75 1.75 0 0010.747 15H11a.75.75 0 000-1.5h-.253a.25.25 0 01-.244-.304l.459-2.066A1.75 1.75 0 009.253 9H9z",
  "clip-rule": "evenodd"
}, null, -1), Re = [
  De
];
function ze(s, t) {
  return l(), h("svg", Ae, Re);
}
const qe = /* @__PURE__ */ H(Pe, [["render", ze]]), He = {}, Le = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 24 24",
  fill: "currentColor",
  height: "20",
  width: "20"
}, Ve = /* @__PURE__ */ $("path", {
  "fill-rule": "evenodd",
  d: "M9.401 3.003c1.155-2 4.043-2 5.197 0l7.355 12.748c1.154 2-.29 4.5-2.599 4.5H4.645c-2.309 0-3.752-2.5-2.598-4.5L9.4 3.003zM12 8.25a.75.75 0 01.75.75v3.75a.75.75 0 01-1.5 0V9a.75.75 0 01.75-.75zm0 8.25a.75.75 0 100-1.5.75.75 0 000 1.5z",
  "clip-rule": "evenodd"
}, null, -1), Fe = [
  Ve
];
function Ue(s, t) {
  return l(), h("svg", Le, Fe);
}
const je = /* @__PURE__ */ H(He, [["render", Ue]]), Ke = {}, We = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 20 20",
  fill: "currentColor",
  height: "20",
  width: "20"
}, Ye = /* @__PURE__ */ $("path", {
  "fill-rule": "evenodd",
  d: "M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-8-5a.75.75 0 01.75.75v4.5a.75.75 0 01-1.5 0v-4.5A.75.75 0 0110 5zm0 10a1 1 0 100-2 1 1 0 000 2z",
  "clip-rule": "evenodd"
}, null, -1), Ge = [
  Ye
];
function Je(s, t) {
  return l(), h("svg", We, Ge);
}
const Qe = /* @__PURE__ */ H(Ke, [["render", Je]]), Xe = {}, Ze = {
  xmlns: "http://www.w3.org/2000/svg",
  width: "12",
  height: "12",
  viewBox: "0 0 24 24",
  fill: "none",
  stroke: "currentColor",
  "stoke-width": "1.5",
  "stroke-linecap": "round",
  "stroke-linejoin": "round"
}, et = /* @__PURE__ */ $("line", {
  x1: "18",
  y1: "6",
  x2: "6",
  y2: "18"
}, null, -1), tt = /* @__PURE__ */ $("line", {
  x1: "6",
  y1: "6",
  x2: "18",
  y2: "18"
}, null, -1), st = [
  et,
  tt
];
function ot(s, t) {
  return l(), h("svg", Ze, st);
}
const it = /* @__PURE__ */ H(Xe, [["render", ot]]), nt = ["aria-live", "data-styled", "data-mounted", "data-promise", "data-removed", "data-visible", "data-y-position", "data-x-position", "data-index", "data-front", "data-swiping", "data-type", "data-invert", "data-swipe-out", "data-expanded"], at = ["data-disabled"], rt = {
  key: 1,
  "data-icon": ""
}, lt = { "data-content": "" }, ut = { "data-title": "" }, ct = 4e3, dt = 14, vt = 20, ft = 200, ht = /* @__PURE__ */ ee({
  __name: "Toast",
  props: {
    toast: {
      type: Object,
      required: !0
    },
    toasts: {
      type: Array,
      required: !0
    },
    index: {
      type: Number,
      required: !0
    },
    expanded: {
      type: Boolean,
      required: !0
    },
    invert: {
      type: Boolean,
      required: !0
    },
    heights: {
      type: Array,
      required: !0
    },
    position: {
      type: String,
      required: !0
    },
    visibleToasts: {
      type: Number,
      required: !0
    },
    expandByDefault: {
      type: Boolean,
      required: !0
    },
    closeButton: {
      type: Boolean,
      required: !0
    },
    interacting: {
      type: Boolean,
      required: !0
    },
    duration: {
      type: Number,
      required: !1
    },
    descriptionClass: {
      type: String,
      required: !1
    }
  },
  emits: ["update:heights", "removeToast"],
  setup(s, { emit: t }) {
    const e = s, i = (a) => !!a.promise, u = c(!1), m = c(!1), d = c(!1), p = c(!1), v = c(null), B = c(0), L = c(0), V = c(null), x = c(null), o = w(() => e.index === 0), n = w(() => e.index + 1 <= e.visibleToasts), f = w(() => e.toast.type), b = w(() => e.toast.dismissible), S = e.toast.className || "", _ = e.toast.descriptionClassName || "", E = e.toast.style || {}, P = w(
      () => e.heights.findIndex((a) => a.toastId === e.toast.id) || 0
    ), F = w(
      () => e.toast.duration || e.duration || ct
    ), A = c(0), D = c(0), W = c(F.value), te = c(0), U = c(null), Y = w(() => e.position.split("-")), le = w(() => e.heights.reduce((a, r, g) => g >= P.value ? a : a + r.height, 0)), ue = w(() => e.toast.invert || e.invert), G = w(() => v.value === "loading"), j = w(
      () => v.value ?? (e.toast.type || null)
    ), J = w(() => !i(e.toast) && typeof e.toast.title == "object"), ce = w(() => {
      if (!i(e.toast))
        return null;
      switch (v.value) {
        case "loading":
          return e.toast.loading;
        case "success":
          return typeof e.toast.success == "function" ? V.value : e.toast.success;
        case "error":
          return typeof e.toast.error == "function" ? V.value : e.toast.error;
        default:
          return null;
      }
    });
    Q(() => u.value = !0), q(() => {
      D.value = P.value * dt + le.value;
    });
    function de() {
      var a, r;
      (!G.value || b.value) && (I(), (r = (a = e.toast).onDismiss) == null || r.call(a, e.toast));
    }
    function I() {
      m.value = !0, B.value = D.value;
      const a = e.heights.filter(
        (r) => r.toastId !== e.toast.id
      );
      t("update:heights", a), setTimeout(() => {
        t("removeToast", e.toast);
      }, ft);
    }
    const ve = (a) => {
      G || (B.value = D.value, a.target.setPointerCapture(a.pointerId), a.target.tagName !== "BUTTON" && (d.value = !0, U.value = a.clientY));
    }, fe = (a) => {
      var g, N, k, R;
      if (p.value)
        return;
      const r = Number(
        ((g = x.value) == null ? void 0 : g.style.getPropertyValue("--swipe-amount").replace("px", "")) || 0
      );
      if (Math.abs(r) >= vt) {
        B.value = D.value, (k = (N = e.toast).onDismiss) == null || k.call(N, e.toast), I(), p.value = !0;
        return;
      }
      (R = x.value) == null || R.style.setProperty("--swipe-amount", "0px"), U.value = null, d.value = !0;
    }, he = (a) => {
      var N, k;
      if (!U.value)
        return;
      const r = a.clientY - U.value;
      if (!(Y.value[0] === "top" ? r < 0 : r > 0)) {
        (N = x.value) == null || N.style.setProperty("--swipe-amount", "0px");
        return;
      }
      (k = x.value) == null || k.style.setProperty("--swipe-amount", `${r}px`);
    };
    return q((a) => {
      if (e.toast.promise && v.value === "loading" || e.toast.duration === 1 / 0)
        return;
      let r;
      const g = () => {
        if (te.value < A.value) {
          const k = (/* @__PURE__ */ new Date()).getTime() - A.value;
          W.value = W.value - k;
        }
        te.value = (/* @__PURE__ */ new Date()).getTime();
      }, N = () => {
        A.value = (/* @__PURE__ */ new Date()).getTime(), r = setTimeout(() => {
          var k, R;
          (R = (k = e.toast).onAutoClose) == null || R.call(k, e.toast), I();
        }, W.value);
      };
      e.expanded || e.interacting ? g() : N(), a(() => {
        clearTimeout(r);
      });
    }), Q(() => {
      if (x.value) {
        const a = x.value.getBoundingClientRect().height;
        L.value = a;
        const r = [{ toastId: e.toast.id, height: a }, ...e.heights];
        t("update:heights", r);
      }
    }), re(() => {
      if (x.value) {
        const a = e.heights.filter(
          (r) => r.toastId !== e.toast.id
        );
        t("update:heights", a);
      }
    }), q(() => {
      e.toast.delete && I();
    }), (a, r) => (l(), h("li", {
      "aria-live": s.toast.important ? "assertive" : "polite",
      "aria-atomic": "",
      role: "status",
      tabindex: "0",
      ref_key: "toastRef",
      ref: x,
      "data-sonner-toast": "",
      class: se(M(S)),
      "data-styled": !J.value,
      "data-mounted": u.value,
      "data-promise": !!s.toast.promise,
      "data-removed": m.value,
      "data-visible": n.value,
      "data-y-position": Y.value[0],
      "data-x-position": Y.value[1],
      "data-index": e.index,
      "data-front": o.value,
      "data-swiping": d.value,
      "data-type": v.value !== "loading" && v.value ? v.value : f.value,
      "data-invert": ue.value,
      "data-swipe-out": p.value,
      "data-expanded": !!(e.expanded || e.expandByDefault && u.value),
      style: X({
        "--index": e.index,
        "--toasts-before": e.index,
        "--z-index": s.toasts.length - e.index,
        "--offset": `${m.value ? B.value : D.value}px`,
        "--initial-height": e.expandByDefault ? "auto" : `${L.value}px`,
        ...M(E)
      }),
      onPointerdown: ve,
      onPointerup: fe,
      onPointermove: he
    }, [
      e.closeButton && !J.value ? (l(), h("button", {
        key: 0,
        "aria-label": "Close toast",
        "data-close-button": "",
        "data-disabled": G.value,
        onClick: de
      }, [
        ge(it)
      ], 8, at)) : C("", !0),
      f.value || s.toast.icon || s.toast.promise ? (l(), h("div", rt, [
        typeof s.toast.promise == "function" || f.value === "loading" ? (l(), O(_e, {
          key: 0,
          visible: v.value === "loading" || f.value === "loading"
        }, null, 8, ["visible"])) : C("", !0),
        j.value === "success" ? (l(), O(Me, { key: 1 })) : j.value === "info" ? (l(), O(qe, { key: 2 })) : j.value === "warning" ? (l(), O(je, { key: 3 })) : j.value === "error" ? (l(), O(Qe, { key: 4 })) : C("", !0)
      ])) : C("", !0),
      $("div", lt, [
        $("div", ut, [
          typeof s.toast.title == "string" ? (l(), h(K, { key: 0 }, [
            oe(z(s.toast.title), 1)
          ], 64)) : s.toast.title === void 0 || s.toast.title === null ? (l(), h(K, { key: 1 }, [
            oe(z(ce.value), 1)
          ], 64)) : J.value ? (l(), O(ye(s.toast.title), {
            key: 2,
            onCloseToast: r[0] || (r[0] = () => {
              var g;
              I(), (g = s.toast.cancel) != null && g.onClick && s.toast.cancel.onClick();
            })
          }, null, 32)) : C("", !0)
        ]),
        s.toast.description ? (l(), h("div", {
          key: 0,
          "data-description": "",
          class: se(s.descriptionClass + M(_))
        }, z(s.toast.description), 3)) : C("", !0)
      ]),
      s.toast.cancel ? (l(), h("button", {
        key: 2,
        "data-button": "",
        "data-cancel": "",
        onClick: r[1] || (r[1] = () => {
          var g;
          I(), (g = s.toast.cancel) != null && g.onClick && s.toast.cancel.onClick();
        })
      }, z(s.toast.cancel.label), 1)) : C("", !0),
      s.toast.action ? (l(), h("button", {
        key: 3,
        "data-button": "",
        onClick: r[2] || (r[2] = () => {
          var g;
          I(), (g = s.toast.action) == null || g.onClick();
        })
      }, z(s.toast.action.label), 1)) : C("", !0)
    ], 46, nt));
  }
});
const mt = ["aria-label"], pt = ["dir", "data-theme", "data-rich-colors", "data-y-position", "data-x-position"], gt = 3, ie = "32px", yt = 4e3, wt = 356, ne = 14, bt = /* @__PURE__ */ ee({
  name: "Toaster",
  inheritAttrs: !1,
  __name: "Toaster",
  props: {
    invert: { type: Boolean, default: !1 },
    theme: { default: "light" },
    position: { default: "bottom-right" },
    hotkey: { default: () => ["altKey", "KeyT"] },
    richColors: { type: Boolean, default: !1 },
    expand: { type: Boolean, default: !1 },
    duration: { default: yt },
    gap: { default: ne },
    visibleToasts: { default: gt },
    closeButton: { type: Boolean, default: !1 },
    toastOptions: { default: () => ({}) },
    className: { default: "" },
    style: { default: () => ({}) },
    offset: { default: ie },
    dir: { default: "auto" }
  },
  setup(s) {
    const t = s, e = we(), i = c([]), u = c([]), m = c(!1), d = c(!1), p = c(
      t.theme !== "system" ? t.theme : typeof window < "u" && window.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light"
    ), v = w(() => t.position.split("-")), B = c(null), L = t.hotkey.join("+").replace(/Key/g, "").replace(/Digit/g, "");
    function V(o) {
      i.value = i.value.filter(({ id: n }) => n !== o.id);
    }
    function x() {
      if (typeof window > "u")
        return "ltr";
      const o = document.documentElement.getAttribute("dir");
      return o === "auto" || !o ? window.getComputedStyle(document.documentElement).direction : o;
    }
    return Q(() => {
      const o = T.subscribe((n) => {
        if (n.dismiss) {
          i.value = i.value.map(
            (f) => f.id === n.id ? { ...f, delete: !0 } : f
          );
          return;
        }
        be(() => {
          const f = i.value.findIndex((b) => b.id === n.id);
          f !== -1 ? i.value.splice(f, 1, n) : i.value = [n, ...i.value];
        });
      });
      re(() => {
        o();
      });
    }), Te(
      () => t.theme,
      (o) => {
        if (o !== "system") {
          p.value = o;
          return;
        }
        o === "system" && (window.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches ? p.value = "dark" : p.value = "light"), !(typeof window > "u") && window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", ({ matches: n }) => {
          n ? p.value = "dark" : p.value = "light";
        });
      }
    ), q(() => {
      i.value.length <= 1 && (m.value = !1);
    }), q((o) => {
      function n(f) {
        var S, _;
        t.hotkey.every(
          (E) => f[E] || f.code === E
        ) && (m.value = !0, (S = B.value) == null || S.focus()), f.code === "Escape" && (document.activeElement === B.value || (_ = B.value) != null && _.contains(document.activeElement)) && (m.value = !1);
      }
      document.addEventListener("keydown", n), o(() => {
        document.removeEventListener("keydown", n);
      });
    }), (o, n) => {
      var f;
      return l(), h("section", {
        "aria-label": `Notifications ${M(L)}`,
        tabIndex: -1
      }, [
        $("ol", {
          ref_key: "listRef",
          ref: B,
          "data-sonner-toaster": "",
          dir: o.dir === "auto" ? x() : o.dir,
          tabIndex: -1,
          "data-theme": o.theme,
          "data-rich-colors": o.richColors,
          "data-y-position": v.value[0],
          "data-x-position": v.value[1],
          style: X(
            {
              "--front-toast-height": `${(f = u.value[0]) == null ? void 0 : f.height}px`,
              "--offset": typeof o.offset == "number" ? `${o.offset}px` : o.offset || ie,
              "--width": `${wt}px`,
              "--gap": `${ne}px`,
              ...M(e).style
            }
          ),
          onMouseenter: n[1] || (n[1] = (b) => m.value = !0),
          onMousemove: n[2] || (n[2] = (b) => m.value = !0),
          onMouseleave: n[3] || (n[3] = () => {
            d.value || (m.value = !1);
          }),
          onPointerdown: n[4] || (n[4] = (b) => d.value = !1),
          onPointerup: n[5] || (n[5] = (b) => d.value = !1)
        }, [
          (l(!0), h(K, null, ae(i.value, (b, S) => {
            var _, E, P, F;
            return l(), O(ht, {
              key: b.id,
              index: S,
              toast: b,
              duration: ((_ = o.toastOptions) == null ? void 0 : _.duration) ?? o.duration,
              className: (E = o.toastOptions) == null ? void 0 : E.className,
              descriptionClassName: (P = o.toastOptions) == null ? void 0 : P.descriptionClassName,
              invert: o.invert,
              visibleToasts: o.visibleToasts,
              closeButton: o.closeButton,
              interacting: d.value,
              position: o.position,
              style: X((F = o.toastOptions) == null ? void 0 : F.style),
              toasts: i.value,
              expandByDefault: o.expand,
              gap: o.gap,
              expanded: m.value,
              heights: u.value,
              "onUpdate:heights": n[0] || (n[0] = (A) => u.value = A),
              onRemoveToast: V
            }, null, 8, ["index", "toast", "duration", "className", "descriptionClassName", "invert", "visibleToasts", "closeButton", "interacting", "position", "style", "toasts", "expandByDefault", "gap", "expanded", "heights"]);
          }), 128))
        ], 44, pt)
      ], 8, mt);
    };
  }
}), Bt = {
  install(s) {
    s.component("Toaster", bt);
  }
};
export {
  bt as Toaster,
  Bt as default,
  $t as toast
};
